### DISCUSSION TITLE
Styling Tooltip

### DISCUSSION THREAD

**Discussion by sylcha** [mdbootstrap/TW-Elements#2225] [Support from community]
Hi everyone!

Is there a way to apply a CSS style (with Tailwind syntax) to a [tooltip](https://tw-elements.com/docs/standard/components/tooltip/) : I wandered within the documentation but found no mention of it (which does not mean that there is not though...).

Thanks in advance for your answers,
Sylvain

**Comment by juujisai** [comment]
Hello. You would have to update the tooltips `template` option to make the changes in tooltip styling. The default template option looks like this (you can add one in a single string, doesn't have to be in 3 different like bellow)

```
  template:
    '<div class="opacity-0 transition-opacity duration-300 ease-in-out absolute z-[1080] block m-0 text-sm not-italic font-normal text-left no-underline underline-offset-auto normal-case leading-6 tracking-normal break-normal whitespace-normal" role="tooltip">' +
    '<div data-te-tooltip-inner-ref class="tooltip-inner max-w-[200px] text-sm py-1.5 px-4 text-white text-center bg-[#6d6d6d] rounded"></div>' +
    "</div>",
```

**Reply by sylcha** [reply]
Thanks for your answer. I've tried to use this `template` option but I was doing it wrong. 

Now it works perfectly except for one thing : it does not take dark mode option into account. At least the way I'm trying to implement it. I've tried for example to add  `dark:text-black` and `dark:bg-[#e6e6e6]` next to the other colors declarations without results.

**Reply by juujisai** [reply]
Did you check whether inside the browsers inspect tool the classes are added? And whether tailwind is creating new styles for those classes? Maybe tailwind does not see classes from your JS file?

**Reply by sylcha** [reply]
The styles for dark mode are declared properly as you can see in this image.
![image](https://github.com/mdbootstrap/TW-Elements/assets/2913672/3bf68ebb-4d39-482a-9a4d-50095e2e77ad)

I use Tailwind CSS along with Svelte and in the most outter div of a component, I declare inside the `class` property : `{$darkMode.isActive ? 'dark'  : ''}` where `darkMode` is a store where I keep track of dark mode state. That set up enable me to use the `dark:...` syntax for every div within the component and it works as expected as you can see in the image below... except for the tooltip which keep displaying the light mode colors!
![image](https://github.com/mdbootstrap/TW-Elements/assets/2913672/7326bb89-fa89-42d5-b9de-4fd6f866c563)

**Reply by juujisai** [reply]
Hard to tell,  I think everything looks fine. Did you see the styles beeing applied with the `dark:bg:coopmathsdark-darkest` on any other components? Maybe tailwind didn't create them. For a test, try adding some random span tag and apply this class to that span and see whether the tooltip picks it up. You can also try to add a class with a background that already exsits on your website and see whether anything changes.

If you click on the tooltip toggler, the tooltip will stay visible until you click something again - you can scroll through the DOM inside your browser inspect tool and at the very bottom there will be the code for the tooltip. Make sure that each of the classes are beeing added there - it should be fine though because I see your tooltip looks different than the default.

Make sure that the file where you are adding the component is beeing included in `tailwind.config.js` - are other classes that were added in this file working properly? Classes that werent present on any other files.

**Reply by sylcha** [reply]
If I inspect *in the same page where the tooltip is defined* an element that is displayed with the good background color in dark mode, the style inspector first item is :
```
:is(.dark .dark\:bg-coopmathsdark-canvas-dark) {
    --tw-bg-opacity: 1;
    background-color: rgb(58 61 78 / var(--tw-bg-opacity));
}
```
I used your trick and add `data-te-trigger="click"` to display the tooltip continuously. The treatment I mentioned above does not appear for the tooltip. However, the tooltip seems to get the right colors declaration for dark mode
![image](https://github.com/mdbootstrap/TW-Elements/assets/2913672/c51ebfb9-7f95-4ce4-8b89-9af1c53b20b0)
so I don't understand what happens here. 

How do you handle within tailwind dark mode for tooltips usually?

**Reply by juujisai** [reply]
This code works for me:

<p>
  Hover the link to see the
  <a
    href="#"
    class="transititext-primary text-primary transition duration-150 ease-in-out hover:text-primary-600 focus:text-primary-600 active:text-primary-700 dark:text-primary-400 dark:hover:text-primary-500 dark:focus:text-primary-500 dark:active:text-primary-600"
    data-te-toggle="tooltip"
    data-te-template='<div class="opacity-0 transition-opacity duration-300 ease-in-out absolute z-[1080] block m-0 text-sm not-italic font-normal text-left no-underline underline-offset-auto normal-case leading-6 tracking-normal break-normal whitespace-normal" role="tooltip">
      <div data-te-tooltip-inner-ref class="tooltip-inner max-w-[200px] text-sm py-1.5 px-4 text-white text-center bg-blue-600 rounded dark:bg-red-600"></div></div>'
    title="Hi! I'm tooltip"
    >tooltip</a
  >
</p>

When I change to the darkmode (added class `dark` to the `html` element) - the tooltip is changing color from blue to red. Does it work for you the same?

![image](https://github.com/mdbootstrap/TW-Elements/assets/52455076/c2483dbf-ea9e-48eb-bac6-40c176566c66)

![image](https://github.com/mdbootstrap/TW-Elements/assets/52455076/5509d2b9-2b33-45c7-a911-94f9283cc610)

**Reply by sylcha** [reply]
If you mean by your question if the tooltip appears with the good color (blue in light mode and red in dark mode) in _this_ page, well, it does not.
Maybe you forgot to write the code in your previous comment ?

**Reply by juujisai** [reply]
Oh wow, Im sorry. Seems like I pasted the wrong text

```
<p>
  Hover the link to see the
  <a
    href="#"
    class="transititext-primary text-primary transition duration-150 ease-in-out hover:text-primary-600 focus:text-primary-600 active:text-primary-700 dark:text-primary-400 dark:hover:text-primary-500 dark:focus:text-primary-500 dark:active:text-primary-600"
    data-te-toggle="tooltip"
    data-te-template='<div class="opacity-0 transition-opacity duration-300 ease-in-out absolute z-[1080] block m-0 text-sm not-italic font-normal text-left no-underline underline-offset-auto normal-case leading-6 tracking-normal break-normal whitespace-normal" role="tooltip">
      <div data-te-tooltip-inner-ref class="tooltip-inner max-w-[200px] text-sm py-1.5 px-4 text-white text-center bg-blue-600 rounded dark:bg-red-600"></div></div>'
    title="Hi! I'm tooltip"
    >tooltip</a
  >
</p>
```

This is the code for tooltip that I've shown on the screens yesterday

**Reply by sylcha** [reply]
Well, in my case, it still blue in dark mode and doesn't turn red.

Another weird thing happens. I use the very same Svelte component (code below) in two different places. In one place, style applies correctly (except for the dark mode colors). Here is what displayed in the inspector :

```
<span class="inline-block" data-te-toggle="tooltip" data-te-placement="bottom" data-te-template="<div class=&quot;opacity-0 transition-opacity duration-300 ease-in-out absolute z-[1080] block m-0 text-sm not-italic font-normal text-left no-underline underline-offset-auto normal-case leading-6 tracking-normal break-normal whitespace-normal&quot; role=&quot;tooltip&quot;><div data-te-tooltip-inner-ref class=&quot;tooltip-inner max-w-[200px] text-sm py-1.5 px-4 text-coopmaths-corpus-light dark:text-coopmathsdark-corpus-light text-center font-extralight bg-coopmaths-canvas-dark dark:bg-coopmathsdark-canvas-dark rounded-sm shadow-lg&quot;></div></div>" title="" data-te-original-title="Réduire la taille du texte" aria-label="Réduire la taille du texte"><button type="button" id="btn-269906712804013136424" class="flex items-center text-3xl text-coopmaths-action dark:text-coopmathsdark-action
        hover:text-coopmaths-action-lightest dark:hover:text-coopmathsdark-action-lightest"><i class="bx bx-zoom-out"></i></button></span>
```

On the other place, it has a different treatment and style does not apply at all :

```
<span class="inline-block" data-te-toggle="tooltip" data-te-placement="bottom" data-te-template="<div class=&quot;opacity-0 transition-opacity duration-300 ease-in-out absolute z-[1080] block m-0 text-sm not-italic font-normal text-left no-underline underline-offset-auto normal-case leading-6 tracking-normal break-normal whitespace-normal&quot; role=&quot;tooltip&quot;><div data-te-tooltip-inner-ref class=&quot;tooltip-inner max-w-[200px] text-sm py-1.5 px-4 text-coopmaths-corpus-light dark:text-coopmathsdark-corpus-light text-center font-extralight bg-coopmaths-canvas-dark dark:bg-coopmathsdark-canvas-dark rounded-sm shadow-lg&quot;></div></div>" title="Réduire la taille du texte"><button type="button" id="btn-3495029182152060655553" class="flex items-center text-3xl text-coopmaths-action dark:text-coopmathsdark-action
        hover:text-coopmaths-action-lightest dark:hover:text-coopmathsdark-action-lightest"><i class="bx bx-zoom-out"></i></button></span>
```

Have you noticed the difference ? In the first case (the one which works as expected), there is `title="" data-te-original-title="Réduire la taille du texte" aria-label="Réduire la taille du texte"` and on the other one (the one which behaves weirdly) just ` title="Réduire la taille du texte"`. I'm not sure whether that's why the second one does not have style applied though.

I don't know why this happens but this inconstancy drives me mad! Do you have any idea why it does not behave the same in the two places?

---

Here is the code of my button svelte component :
```html
<script lang="ts">
  import { onMount } from 'svelte'
  import { getUniqueStringBasedOnTimeStamp } from '../utils/time'
  import { Tooltip, initTE } from 'tw-elements'
  import { tooltipTemplate } from '../../lib/format/tooltips'
  export let title: string = 'Valider'
  export let isDisabled: boolean = false
  export let icon: string = ''
  export let idLabel: string = getUniqueStringBasedOnTimeStamp('btn-')
  export let inverted: boolean = false
  export let tooltipTitle: string = 'Mon tooltip'
  export let tooltipPlacement: 'auto' | 'top' | 'bottom' | 'left' | 'right' =
    'top'

  onMount(() => {
    initTE({ Tooltip })
  })
</script>

<span
  class="inline-block"
  data-te-toggle="tooltip"
  data-te-placement={tooltipPlacement}
  data-te-template={tooltipTemplate}
  title={tooltipTitle}
>
  {#if title.length === 0}
    <button
      type="button"
      id={idLabel}
      class={`${
        $$props.class || ''
      } text-coopmaths-action dark:text-coopmathsdark-action
        ${
          isDisabled
            ? ' text-opacity-10 dark:text-opacity-10'
            : 'hover:text-coopmaths-action-lightest dark:hover:text-coopmathsdark-action-lightest'
        }`}
      disabled={isDisabled}
      on:click
    >
      <i class="bx {icon}" />
    </button>
  {:else}
    <button
      type="button"
      id={idLabel}
      class={`${$$props.class || ''}
    ${
      inverted
        ? 'text-coopmaths-action dark:text-coopmathsdark-action bg-coopmaths-canvas dark:bg-coopmathsdark-canvas border border-coopmaths-action hover:bg-coopmaths-action dark:hover:bg-coopmathsdark-action hover:text-coopmaths-canvas dark:hover:text-coopmathsdark-canvas'
        : 'text-coopmaths-canvas dark:text-coopmathsdark-canvas bg-coopmaths-action dark:bg-coopmathsdark-action hover:bg-coopmaths-action-lightest dark:hover:bg-coopmathsdark-action-lightest'
    }
      ${
        isDisabled
          ? 'dark:bg-coopmathsdark-action bg-opacity-10 dark:bg-opacity-10'
          : ''
      }`}
      disabled={isDisabled}
      on:click
    >
      {#if icon !== ''}
        <i class="bx {icon}" />
      {/if}
      {title}
    </button>
  {/if}
</span>
```

and I use it like this for exemple :

```html
<ButtonWithTooltip
  tooltipTitle="Réduire la taille du texte"
  tooltipPlacement="bottom"
  title=""
  icon="bx-zoom-in"
  class="flex items-center text-3xl"
  on:click={zoomPlus}
/>
```

**Reply by juujisai** [reply]
Seems like the second one is not beeing initialized. You can try moving the `initTE` method to a parent component to both of those buttons or if thats not possible / wont work, try to initialize the component via JS (instead of initTE).

```
new Tooltip(document.querySelector("#tooltip-1"));
```

I do not see the reason why the darkmode classes doesn't work for you. Are you using tailwinds intelisense? Does it show styles of the tailwind classes after hover?
